src/nets/me2e_asr_attctc.py

```diff
@@ -0,0 +1,204 @@
+                        for ch in utt] for utt in xs_real]
+            hs_imag = [[chainer.Variable(self.xp.array(ch, dtype=np.float32))
+                        for ch in utt] for utt in xs_imag]
+
+            hs = {}
+            hs['real'] = hs_real
+            hs['imag'] = hs_imag
+
+            # 1. beamformer
+            hs = self.enhan(hs)
+        else:
+            logging.error(
+                "Error: need to specify an appropriate training mode")
+            sys.exit()
+
+        # utt list of olen
+        ys = [self.xp.array(
+            list(map(int, i[1]['tokenid'].split())), dtype=np.int32) for i in data]
+        ys = [chainer.Variable(y) for y in ys]
+
+        # 2. feature extractor
+        hs = self.logmel(hs, self.xp)
+        hs = self.cmvn(hs, self.xp)
+
+        # 3. encoder
+        hs, ilens = self.asr.enc(hs, ilens)
+
+        # 4. CTC loss
+        loss_ctc = self.asr.ctc(hs, ys)
+
+        # 5. attention loss
+        loss_att, acc, att_w = self.asr.dec(hs, ys)
+
+        # get alignment
+        '''
+        if self.verbose > 0 and self.outdir is not None:
+            for i in six.moves.range(len(data)):
+                utt = data[i][0]
+                align_file = self.outdir + '/' + utt + '.ali'
+                with open(align_file, "w") as f:
+                    logging.info('writing an alignment file to' + align_file)
+                    pickle.dump((utt, att_w[i]), f)
+        '''
+
+        return loss_ctc, loss_att, acc
+
+    def recognize(self, x, recog_args, char_list, rnnlm=None):
+        '''E2E greedy/beam search
+
+        :param x:
+        :param recog_args:
+        :param char_list:
+        :return:
+        '''
+        logging.info('mode: ' + recog_args.mode)
+
+        if recog_args.mode == 'noisy':
+            x_real = x['real']
+            x_imag = x['imag']
+            ilen = self.xp.array(x_real.shape[0], dtype=np.int32)
+            h_real = chainer.Variable(self.xp.array(x_real, dtype=np.float32))
+            h_imag = chainer.Variable(self.xp.array(x_imag, dtype=np.float32))
+
+            # make a utt list (1) to use the same interface for encoder
+            hs = {}
+            hs['real'] = [h_real]
+            hs['imag'] = [h_imag]
+        elif recog_args.mode == 'enhan':
+            x_real = x['real']
+            x_imag = x['imag']
+            ilen = self.xp.array(x_real[0].shape[0], dtype=np.int32)
+            h_real = [chainer.Variable(self.xp.array(ch, dtype=np.float32)) for ch in x_real]
+            h_imag = [chainer.Variable(self.xp.array(ch, dtype=np.float32)) for ch in x_imag]
+
+            # make a utt list (1) to use the same interface for encoder
+            hs = {}
+            hs['real'] = [h_real]
+            hs['imag'] = [h_imag]
+
+            # 1. beamformer
+            hs = self.enhan(hs)
+        else:
+            logging.error(
+                "Error: need to specify an appropriate training mode")
```

Comment:
use `raise NotImplementedError("unknown training mode %s" % recog_args.mode)` not to go on.